jupyter_server_proxy/handlers.py

```diff
@@ -457,6 +460,32 @@ def get_timeout(self):
                 return False
 
     async def ensure_process(self):
+        """
+        Start the process
+        """
+        # The default asyncio event loop implementation on Windows does not 
+        # support subprocesses. Subprocesses are available for Windows if a 
+        # ProactorEventLoop is used. 
+        # Ref: https://docs.python.org/3/library/asyncio-subprocess.html
+        # We use a workaround for windows by starting a new ProactorEventLoop
+        # in a thread. This is necessary because we don't want it to interfere with
+        # the event loop for tornado web server (which does not support ProactorEventLoop).
+        if sys.platform == "win32":
+            def start_thread():
+                new_loop = asyncio.ProactorEventLoop()
+                # patch add_signal_handler for simpervisor
+                def add_signal_handler(signum, callback, *args):
+                    pass
+                new_loop.add_signal_handler = add_signal_handler
+                asyncio.set_event_loop(new_loop)
+                new_loop.run_until_complete(self._ensure_process())
+                new_loop.close()
+            loop = asyncio.get_event_loop()
+            await loop.run_in_executor(None, start_thread)
+        else:
+            await self._ensure_process()
+
+    async def _ensure_process(self):
```

Comment:
`_ensure_process` can now be called either from the main thread (where all the other couroutines are called from), or from an entirely different, ephemeral thread with its own event loop (which will only call _ensure_process). This might need some amount of cross-thread synchronization - for example, `self.state` could now be called from multiple threads. Does this sound correct?

I'm not actually sure how to go from here - my experience with python multi-threading is very limited, and mixing multi-threading with two event loops is even more limited.

Reply:
In our particular case, the `self.state['proc_lock']` and `self.state['proc']` is only used in the ensure_process thread, we will be completely fine. It will be an issue if multiple thread starts to write it, and we will need to lock the state.